### ISSUE TITLE
Unable to create an instance

### ISSUE THREAD

**Issue by themisir** [beringresearch/macpine#117]
Whenever I try to create an instance I am getting the following error message:

```
qemu-system-x86_64: Property 'host-x86_64-cpu.pdpe1gb=on' not found
```

I also tried to use `-a aarch64` but this time I got ssh handshake failed error. I rather prefer to use host CPU (x86_64) though anyways. Do I have to enable `cpu.pdpe1gb` feature somehow or there's something else that's causing the issue. Any pointers would be appreciated!

My host machine is: MacBook Pro 2019 (Intel model)

**Comment by idroz** [comment]
Hi- thanks for raising this. Could you check your qemu version number?

**Comment by themisir** [comment]
> Hi- thanks for raising this. Could you check your qemu version number?

Sure, here you go:

```sh
❯ qemu-system-x86_64 --version
QEMU emulator version 7.2.1
Copyright (c) 2003-2022 Fabrice Bellard and the QEMU Project developers

❯ brew upgrade qemu
==> Downloading https://formulae.brew.sh/api/formula.jws.json
######################################################################## 100.0%
==> Downloading https://formulae.brew.sh/api/cask.jws.json
######################################################################## 100.0%
Warning: qemu 7.2.1 already installed
```

**Comment by maxzinkus** [comment]
Yep, that's definitely up to date. Interesting! I don't personally have an Intel mac to test on. Some intel architectures don't support 1GB pages -- I wonder if qemu makes that show up as this "property not found" error.

The ssh handshake error could be related to qemu not parsing or handling remaining options and therefore not opening the port for SSH forwarding.

**Comment by idroz** [comment]
Thanks! I’ll be able to take a look at this, but will be early-ish next week.

I suspect this may be due to a potential bug in how qemu cpuType string is constructed for host cpus using x86 (https://github.com/beringresearch/macpine/blob/d65fa84cc55c7ada7dbcc5f37ee97172a65b2354/qemu/ops.go#L282)

But will investigate in greater detail.

**Comment by themisir** [comment]
> Thanks! I’ll be able to take a look at this, but will be early-ish next week.
> 
> I suspect this may be due to a potential bug in how qemu cpuType string is constructed for host cpus using x86 (
> 
> https://github.com/beringresearch/macpine/blob/d65fa84cc55c7ada7dbcc5f37ee97172a65b2354/qemu/ops.go#L282
> 
> )
> But will investigate in greater detail.

I played a bit with those options.

I tried reverting changes been made using those commits:

1. 76f05e46dabd872d10fa5bc3f2444f21eed4d785
2. 8129af32b58e9baa7129d7e4a7971a5976cb9818

It worked with both `host,-pdpe1gb` and `host` values, but not with the `host,-pdpe1gb=on` value which is the current state.

I also tried with the `host,pdpe1gb=on` value, which also worked.

| Flags | Works? |
|--|--|
| `host,-pdpe1gb=on` | No |
| `host` | Yes |
| `host,pdpe1gb` | Yes |
| `host,pdpe1gb=on` | Yes |

In conclusion, I am not sure how this change may affect the current user base. Also, I am not that good with qemu stuff. So, I am leaving to you guys to choose what next needs to be done.

**Comment by themisir** [comment]
> Same question for the other working combinations in the table?

I updated the table with regarding `pdpe1gb` support information for each configuration. Looks like unless explicitly disabled with `pdpe1gb=off` flag it's enabled by default.

**Comment by idroz** [comment]
@themisir I've created a pull request #123 . Would you be able to test?

It looks like somewhere along the line `"x86_64_host":  "host,-pdpe1gb"` was overwritten with `"x86_64_host":  "host,-pdpe1gb=on",` which is clearly not right :)

I've reverted this to explicitly say `"host,pdpe1gb=off"`

Intel host CPUs on Macs do not seem to support 1gb pages. See #63 for more info on this.

**Comment by maxzinkus** [comment]
@idroz Per [the table above](https://github.com/beringresearch/macpine/issues/117#issuecomment-1502436954) it looks like maybe @themisir's modern (2019 macbook pro) intel machine can run guests with 1GB page support? Possibly only older intel cpus can't?

1GB pages can be a significant optimization for VMs as they spend less time in emulated VMM, depending on their memory usage. If it doesn't need to be disabled it'd be ideal to have it on when available.

Is there a way we can determine if the x86_64 host supports 1GB pages for VMs, and use it if it does?

**Comment by idroz** [comment]
@maxzinkus looking into it now. 

@themisir could you run the following on your host to check if it supports supper pages (mac's version of huge pages)

`echo "#include <mach/vm_statistics.h>" | gcc -dM -E - | grep VM_FLAGS_SUPERPAGE_SIZE_`

**Comment by themisir** [comment]
### Host

```
❯ echo "#include <mach/vm_statistics.h>" | gcc -dM -E - | grep VM_FLAGS_SUPERPAGE_SIZE_
#define VM_FLAGS_SUPERPAGE_SIZE_2MB (SUPERPAGE_SIZE_2MB<<VM_FLAGS_SUPERPAGE_SHIFT)
#define VM_FLAGS_SUPERPAGE_SIZE_ANY (SUPERPAGE_SIZE_ANY << VM_FLAGS_SUPERPAGE_SHIFT)
```

### VM

```
localhost:~# cat /proc/cpuinfo | grep -i pdpe1gb
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology cpuid pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch pti fsgsbase bmi1 avx2 smep bmi2 erms rdseed adx clflushopt xsaveopt xgetbv1 arat
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology cpuid pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch pti fsgsbase bmi1 avx2 smep bmi2 erms rdseed adx clflushopt xsaveopt xgetbv1 arat
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology cpuid pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch pti fsgsbase bmi1 avx2 smep bmi2 erms rdseed adx clflushopt xsaveopt xgetbv1 arat
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology cpuid pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch pti fsgsbase bmi1 avx2 smep bmi2 erms rdseed adx clflushopt xsaveopt xgetbv1 arat
```

**Comment by themisir** [comment]
What about using `sysctl -a` to determine if the system supports 1 GB pages then setting the flag accordingly?

```shell
❯ sysctl -a | grep machdep.cpu.extfeatures
machdep.cpu.extfeatures: SYSCALL XD 1GBPAGE EM64T LAHF LZCNT PREFETCHW RDTSCP TSCI
```

Also, when I tested with not setting `pdpe1gb` flag the VM was created with 1gb pages support. I am not 100% sure, but what if qemu enables it by default if the device supports 1gb pages?

**Comment by idroz** [comment]
> Also, when I tested without setting pdpe1gb flag, the VM was created with 1gb pages support.

This is really odd because according to [QEMU docs on pdpe1gb](https://qemu-project.gitlab.io/qemu/system/qemu-cpu-models.html)

```
Not included by default in any Intel CPU model.
Should be explicitly turned on for all Intel CPU models.
```

`sysctl` could work. I'll try this option with `syscall.Sysctl` from the standard library.

**Comment by idroz** [comment]
Unless, QEMU's `host` CPU just passes through all CPU features without the need to explicitly set flags...?

**Comment by maxzinkus** [comment]
I think the qemu docs mean that the flag is not enabled by default, not that the technology is unavailable.

It looks like it does based on the table posted above! Just gotta make sure `pdpe1gb=off` happens by default on old Intel machines.

**Comment by themisir** [comment]
> Unless, QEMU's `host` CPU just passes through all CPU features without the need to explicitly set flags...?

I'm not sure if this is related but, https://manpages.ubuntu.com/manpages/kinetic/en/man7/qemu-cpu-models.7.html

>  Libvirt supports a third way to configure CPU models known as "Host model".  This uses the QEMU  "Named  model"  feature,  automatically picking a CPU model that is similar the host CPU, and then adding extra features to approximate the host model as closely as  possible. This  does  not guarantee the CPU family, stepping, etc will precisely match the host CPU, as they would with "Host passthrough", but gives much of the benefit of passthrough, while making live migration safe.

I couldn't find qemu documentation regarding `host` cpu model, though some other resources also pointed out that by default, the `host` model tries to approximate the host cpu.

**Comment by maxzinkus** [comment]
Reviewed and +1, but yeah same limitation.
